### ISSUE TITLE
TransformedDistribution broken in MCMC?

### ISSUE THREAD

**Issue by vanAmsterdam** [pyro-ppl/numpyro#663]
A NegativeHalfNormal distribution, defined with an AffineTransform gives negative samples when sampling directly, but when used in mcmc context, only positive samples. This works properly in the pip-version of numpyro 0.2.4 but not in the development version

```python

from jax import numpy as jnp, random
import numpyro
from numpyro.distributions import constraints, Normal, HalfNormal, TransformedDistribution
from numpyro.distributions.transforms import AffineTransform
from numpyro.infer.mcmc import NUTS, MCMC

class NegativeHalfNormal(TransformedDistribution):
    arg_constraints = {'scale': constraints.positive}

    def __init__(self, scale=1., validate_args=None):
        base_dist = HalfNormal(scale)
        self.scale = base_dist.scale
        super(NegativeHalfNormal, self).__init__(base_dist, AffineTransform(0, -1), validate_args=validate_args)

    @property
    def mean(self):
        return -self.base_dist.mean

    @property
    def variance(self):
        return self.base_dist.variance

NegativeHalfNormal().variance

simkey = random.PRNGKey(0)
X      = Normal().sample(simkey, (100,)) - 0.25

def model(X):
    mu = numpyro.sample('mu', NegativeHalfNormal(2.5))
    with numpyro.plate('obs', X.shape[0]):
        numpyro.sample('X', Normal(mu, 1), obs=X)

smps = NegativeHalfNormal().sample(random.PRNGKey(0), (100,))
print(smps.mean()) # -0.748

mcmc_key = random.PRNGKey(12345)
kernel = NUTS(model)
mcmc   = MCMC(kernel, num_warmup=250, num_samples=750)
mcmc.run(mcmc_key, X)
mcmc.print_summary() # mu = 0.04

```

**Comment by fehiepsi** [comment]
@vanAmsterdam Currently, AffineTransform only works for a positive `scale`. What NumPyro 0.24 does can be translated to the code
```python
def model(X):
    from numpyro.infer.reparam import TransformReparam
    with numpyro.handlers.reparam(config={'mu': TransformReparam()}):
        mu = numpyro.sample('mu', NegativeHalfNormal(2.5))

mcmc.print_summary(exclude_deterministic=False)
```
in the `master` branch (see the eight school example in README). Does it serve your purpose?

If you want to use `NegativeHalfNormal` directly, I think we need to provide `support = constraints.less_than(0)` and its corresponding transform but currently, both Pyro and NumPyro do not provide this constraint.

**Comment by vanAmsterdam** [comment]
thanks for quick reply; It did use to work in the previous versions though

This works. However, I got ~500 more diverging transitions for my model and somewhat worse mixing between chains. The model is a bit too involved to create into a minimal working example. Can this be related to the `reparam` functioning? Or are there other important recent changes to mcmc?

**Comment by fehiepsi** [comment]
> worse mixing between chains

Oops, sorry about that.............

AFAIK, we only change [find_heuristic_step_size](http://num.pyro.ai/en/latest/mcmc.html#numpyro.infer.mcmc.HMC) from `True` to `False` to improve compiling time. Could you change it back to `True` to see if that is the reason?

Could you also help us diagnose the issue by trying different random seeds (using `mcmc.run(PRNGKey(new_seed))`)? We haven't noticed regressions yet.

> Can this be related to the reparam functioning?

I think using `reparam` makes two models (`numpyro==0.2.4 + no reparam` and `numpyro==master + reparam`) equivalent. Are there other transformed distributions in your model?

**Comment by fehiepsi** [comment]
@vanAmsterdam Any news for this issue? I think this is a random seed issue but I am not sure... Probably we made some mistakes during refactoring work.

**Comment by vanAmsterdam** [comment]
I found out the divergences are most probably due to a bug in my own code. I have to prioritize running experiments over developing code for now so I can't debug it right now but don't leave this hanging on my account

**Comment by fehiepsi** [comment]
Thanks for the feedbacks @vanAmsterdam! They are really valuable for improving NumPyro. If you observe any regression, please let us know, we'll follow up.

**Comment by vanAmsterdam** [comment]
just for completeness, I can confirm that the extra divergences I found were due to a bug in my own coding. It was due to a numeric overflow / underflow issue that was swept under the carpet by `lax._safe_mul` but is now exposed since jax 0.1.62 (https://jax.readthedocs.io/en/latest/CHANGELOG.html#jax-0-1-62-march-21-2020)
